### PR TITLE
IA-632 leaflet

### PR DESCRIPTION
**Pull request by beygorghor**
Upgrade leaflet (1.9.3) and react-leaflet(3.2.5).

Related JIRA tickets : IA-632

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [x] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- Remove map reducer 
- upgrade every maps to support new versions of packages
- rewrite of custom zoom bar
- provide some docs for frontend and especially for maps
- use only one tile layer component

## How to test

Run `npm ci` first.

Go trough every maps in Iaso and check that everything is ok:
- GeoJsonMapComponent => visible in org unit history for an org unit with a shape
- MarkerMapComponent => visible in instance details with a geo location
- AssignmentsMap => planning
- ListMap => entities /beneficiaries
- InstancesMap => instances list
- OrgUnitsMap => org unit list
- OrgUnitMap => org unit detail
- OrgUnitChildrenMap => registry
- CalendarMap => Polio/calendar

## Notes

Bluesquare component as been upgraded [here](https://github.com/BLSQ/bluesquare-components/pull/121)

### CODE REVIEW

File: hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx [hunk 1]
```diff
@@ -28,52 +34,56 @@ type Props = {
...
+        padding: [10, 10],
+        maxZoom: currentTile.maxZoom,
     };
 
+    const ref: any = useRef(null);
+    useEffect(() => {
+        if (ref.current) {
+            ref.current.setUrl(currentTile.url);
+        }
+    }, [currentTile]);
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx]: I don't understand this part. It seems to me this effect will never run, as `ref.current` is initialized with `null` and is not set anywhere else in the component

File: hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx [hunk 2]
```diff
@@ -28,52 +34,56 @@ type Props = {
...
     }, [geoJson]);
     const [currentTile, setCurrentTile] = useState<Tile>(tiles.osm);
-    const boundsOptions = { padding: [10, 10] };
-
-    const fitToBounds = () => {
-        map.current.leafletElement.fitBounds(bounds, {
-            maxZoom: tiles.osm.maxZoom,
-            padding: boundsOptions.padding,
-        });
+    const boundsOptions: Record<string, any> = {
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx:38]: memo

File: hat/assets/js/apps/Iaso/components/maps/tools/CustomTileLayer.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import React, { FunctionComponent, useRef, useEffect } from 'react';
+import { TileLayer, useMap } from 'react-leaflet';
+import { Tile } from './TilesSwitchDialog';
+
+type Props = {
+    currentTile: Tile;
+};
+
+export const CustomTileLayer: FunctionComponent<Props> = ({ currentTile }) => {
+    const ref: any = useRef(null);
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/tools/CustomTileLayer.tsx]: `RefObject<any>`

File: hat/assets/js/apps/Iaso/components/maps/tools/CustomZoomControl.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        }
+    }, [fitOnLoad, bounds, mapFitted, fitToBounds, map]);
+
+    useEffect(() => {
+        const container = context.layerContainer || context.map;
+
+        const control = L.control.zoom({
+            zoomBoxTitle: formatMessage(MESSAGES['box-zoom-title']),
+            zoomInfoTitle: formatMessage(MESSAGES['info-zoom-title']),
+            fitToBoundsTitle: formatMessage(MESSAGES['fit-to-bounds']),
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/tools/CustomZoomControl.tsx]: Why not `MESSAGES.boxZoomTitle` etc? I see no apparent reason not to use camelCase for the message keys

File: hat/assets/js/apps/Iaso/components/maps/tools/CustomZoomControl.tsx [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    const [mapFitted, setMapFitted] = useState<boolean>(false);
+    const { formatMessage } = useSafeIntl();
+    const fitToBounds = useCallback(() => {
+        if (bounds) {
+            map.fitBounds(bounds, boundsOptions);
+        }
+    }, [bounds, boundsOptions, map]);
+    const context = useLeafletContext();
+
+    L.control.zoom = opts => {
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/tools/CustomZoomControl.tsx]: `opts` could be typed

File: hat/assets/js/apps/Iaso/components/maps/tools/MapToggleCluster.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        fontSize: 12,
+    },
+    box: {
+        backgroundColor: 'white',
+    },
+}));
+
+type Props = {
+    isClusterActive: boolean;
+    setIsClusterActive: Dispatch<SetStateAction<boolean>>;
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/tools/MapToggleCluster.tsx:29]: it just occurred to me that when we want to pass state+setState as props, we could just do 
```javascript
const activeClusterState = useState<boolean>()
// some code
<MapToggleCluster clusterState = {activeClusterState}/>
```

Then in `MapToggleCluster`:
```javascript
const [isClusterActive, setIsClusterActive] = clusterState
```

Not sure we gain much by doing that though, and the linter may complain that the setter should be wrapped in useEffect. 

I just don't like passing these values separately, but passing them together and pattern matching them in the child is not really better ☹️

File: hat/assets/js/apps/Iaso/components/maps/tools/TileSwitchComponent.js
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/tools/TileSwitchComponent.js:1]: As far as I can see, this component is only used by `TileSwitchDialog`, so we could remove `injectIntl` and `withStyles` even if we don't convert it to arrow function without too much risk. Bonus points for TS conversion

File: hat/assets/js/apps/Iaso/domains/instances/components/InstancesMap/useResetMapReducerOnUnmount.ts
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/instances/components/InstancesMap/useResetMapReducerOnUnmount.ts:1]: 🍾

File: hat/assets/js/apps/Iaso/domains/orgUnits/components/orgUnitMap/OrgUnitMap/useRedux.ts
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/components/orgUnitMap/OrgUnitMap/useRedux.ts:1]: Only one left...

File: hat/assets/js/apps/Iaso/components/maps/MarkerMapComponent.tsx [hunk 7]
```diff
@@ -41,61 +34,49 @@ export const MarkerMap: FunctionComponent<Props> = ({
 }) => {
     const [currentTile, setCurrentTile] = useState<Tile>(tiles.osm);
 
-    const map: any = useRef();
-
     const classes: Record<string, string> = useStyles();
 
-    const boundsOptions = { padding: [500, 500] };
+    const boundsOptions: Record<string, any> = {
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/components/maps/MarkerMapComponent.tsx]: `type BoundOptions = Record<string, [number,number] | number>`

File: hat/assets/js/apps/Iaso/domains/entities/components/ListMap.tsx [hunk 8]
```diff
@@ -67,63 +65,47 @@ export const ListMap: FunctionComponent<Props> = ({
     isFetchingLocations,
     extraColumns,
 }) => {
-    const mapContainer: any = useRef();
-    const map: any = useRef();
     const classes: Record<string, string> = useStyles();
     const theme = useTheme();
 
+    const bounds = useMemo(
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/entities/components/ListMap.tsx]: const bounds : Bounds | undefined ?

File: hat/assets/js/apps/Iaso/domains/orgUnits/components/OrgUnitsMap.tsx [hunk 9]
```diff
@@ -104,15 +83,15 @@ const getFullOrgUnits = orgUnits => {
...
+const getOrgUnitsBounds = (orgUnits: Locations): Bounds | undefined => {
     const orgUnitsLocations = getFullOrgUnits(orgUnits.locations);
     const locationsBounds =
         orgUnitsLocations.length > 0
             ? getLatLngBounds(orgUnitsLocations)
             : null;
     const shapeBounds =
         orgUnits.shapes.length > 0 ? getShapesBounds(orgUnits.shapes) : null;
-    let bounds = null;
+    let bounds: Bounds | undefined;
```
Comment (quang-le) [inline hat/assets/js/apps/Iaso/domains/orgUnits/components/OrgUnitsMap.tsx]: nitpick, but style wise,  I'd prefer:
```javascript
if (locationsBounds && shapeBounds) {
    return locationsBounds.extend(shapeBounds);
    } else if (locationsBounds) {
       return locationsBounds;
    } else if (shapeBounds) {
       return shapeBounds;
    }
return undefined
```

**Review by quang-le — CHANGES_REQUESTED** [review]
THE FIT TO BOUNDS NIGHTMARE IS FINALLY OVER!

Nothing really deep to say about the code.

- There's a z-index bug with the tooltips:

Screenshot 2023-05-31 at 14 46 39

- In registry: it's impossible to close the pop-up

https://github.com/BLSQ/iaso/assets/38907762/9027332a-c4e0-4451-bf54-8e241ec0a176

- In org units details map, the fit to bounds seems a bit off:


https://github.com/BLSQ/iaso/assets/38907762/ddcb608e-3744-4e4f-96b9-68f03f454886

File: hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx (hunk 2 above)
Comment (beygorghor) [reply hat/assets/js/apps/Iaso/components/maps/GeoJsonMapComponent.tsx:44]: in fact this is useless, props passed to the map are immutable, to update a value you need to use a hook or leaflet...

File: hat/assets/js/apps/Iaso/components/maps/tools/MapToggleCluster.tsx (hunk 6 above)
Comment (beygorghor) [reply hat/assets/js/apps/Iaso/components/maps/tools/MapToggleCluster.tsx:29]: the isClusterActive is used in the parent component too sometimes ... i'll leave it like that for now

**Comment by beygorghor** [comment]
@Son, regarding the last video, can you check that the parent has a shape and is very small ? i have the same problem on my side with this org unit

**Comment by quang-le** [comment]
> @son, regarding the last video, can you check that the parent has a shape and is very small ? i have the same problem on my side with this org unit

Yes, indeed. I tried an org unit with no parent and the fit to bounds was ok

**Review by quang-le — APPROVED** [review]
LET'S MERGE!!

**Pull request merged**